modules/debugging/index.qmd

```diff
@@ -95,29 +94,271 @@ detach_package(tidyverse)
+-   Don't be disheartened if your error, which you've struggled with for hours, is solved in minutes — we've had plenty of practice (also being stuck for hours) troubleshooting these errors.
+
+# Minimal Reproducible Examples
+
+::: callout-note
+## Remember
+
+A minimally reproducible example (often called an MRE) contains all of the code necessary for someone else to encounter the same error you're facing.
+:::
+
+To create one, you need to provide the **exact packages you're using**, some **data** if needed, and **your code.** A good example is also **minimal**, which means that you've taken the time to remove as much unrelated packages, code and data as possible to produce the error.
+
+## Providing minimal datasets
+
+If your question involves using a dataset, you will need to provide a small dataset to demonstrate the problem. Often people say that the data is confidential and can’t be provided, but usually the problem isn’t specific to the data you are using. Taking the time to try your code on a different dataset is a good debugging step, and R provides many good datasets to use in these examples built in. You can find a complete list of datasets available to all R users with the `data(package = "datasets")` function.
+
+If you do need to use a custom dataset specific to your project, try reducing the size of the dataset before making your example. The error may only occur for some specific rows of your data. Once you've narrowed in on the issue, you can convert your minimal data object to code for producing it using the `dput()` function.
+
+For example you can use use the below code to provide the first 6 rows of the iris dataset.
+
+```{r}
+dput(head(iris))
+```
+
+Try it yourself! Create a minimal dataset using the `dput()` function for the `penguins` dataset.
+
+::: callout-tip
+## Hint
+
+We want to be putting the `penguins` data to create a minimal dataset.
+:::
+
+```{webr-teachr}
+## REVISIT THIS :( 
+# Load the penguins dataset
+library(palmerpenguins)
+# Use dput() to create a minimal dataset
+<<dput(head(penguins))>>
   
 ???
   
+"Did you use dput() with head(penguins)?" != search_ast(.code, .expr = dput(head(penguins)))
+
 ```
 
+You can then copy this generated code, and make it part of your example to create your example data.
+
+Alternatively you can save this minimal dataset to a file, and attach the file in your question.
+
+When you read that file in, make sure you avoid absolute paths to it. Not everyone has the same file structure as you, so they won't be able to access your dataset if it is stored at `/home/learnr/Documents/My Job/project_2040/data/example.csv`. Instead provide a short relative oath, such as `example.csv` or `data/example.csv`. If you're uploading files, it's also nice to provide a zip that contains a R script of your minimal reproducible code, along with the data in the folders referenced by the script.
+
+-- potentially a multiple choice question here on relative paths? ---
```

Comment:
Yes, that'd be great. Which of the following are relative paths?
Include `"~/data/example.csv"` as `~` often confuses/surprises people!

A more realistic name instead of `example.csv` is for the best, something like `housing_prices.csv` for example. I think we can also attach this dataset into the webr chunk if you wanted an interactive example, but a quiz is simpler (and probably better)